app/test/e2e/network-interface-create.e2e.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Copyright Oxide Computer Company
+ */
+import { test } from '@playwright/test'
+
+import { expect } from './utils'
+
+test('can create a NIC with a specified IP address', async ({ page }) => {
+  // go to an instance’s Network Interfaces page
+  await page.goto('/projects/mock-project/instances/db1/network-interfaces')
+
+  // stop the instance
+  await page.getByRole('button', { name: 'Instance actions' }).click()
+  await page.getByRole('menuitem', { name: 'Stop' }).click()
+
+  // open the add network interface side modal
+  await page.getByRole('button', { name: 'Add network interface' }).click()
+
+  // fill out the form
+  await page.getByLabel('Name').fill('nic-1')
+  await page.getByRole('button', { name: 'VPC' }).click()
+  await page.getByRole('option', { name: 'mock-vpc' }).click()
+  await page.getByRole('button', { name: 'Subnet' }).click()
+  await page.getByRole('option', { name: 'mock-subnet' }).click()
+  await page.getByLabel('IP Address').fill('1.2.3.4')
+
+  const sidebar = page.getByRole('dialog', { name: 'Add network interface' })
+
+  // test that the form can be submitted and a new network interface is created
+  await sidebar.getByRole('button', { name: 'Add network interface' }).click()
+  const lastRow = page.getByRole('table').locator('tr:last-child')
+
+  await expect(lastRow.getByText('1.2.3.4')).toBeVisible()
+})
+
+test('can create a NIC with a blank IP address', async ({ page }) => {
+  // go to an instance’s Network Interfaces page
+  await page.goto('/projects/mock-project/instances/db1/network-interfaces')
+
+  // stop the instance
+  await page.getByRole('button', { name: 'Instance actions' }).click()
+  await page.getByRole('menuitem', { name: 'Stop' }).click()
+
+  // open the add network interface side modal
+  await page.getByRole('button', { name: 'Add network interface' }).click()
+
+  // fill out the form
+  await page.getByLabel('Name').fill('nic-2')
+  await page.getByRole('button', { name: 'VPC' }).click()
+  await page.getByRole('option', { name: 'mock-vpc' }).click()
+  await page.getByRole('button', { name: 'Subnet' }).click()
+  await page.getByRole('option', { name: 'mock-subnet' }).click()
+
+  // make sure the IP address field has a non-conforming bit of text in it
+  await page.getByLabel('IP Address').fill('x')
+
+  // try to submit it
+  const sidebar = page.getByRole('dialog', { name: 'Add network interface' })
+  await sidebar.getByRole('button', { name: 'Add network interface' }).click()
+
+  // it should error out
+  // todo: improve error message from API
+  await expect(sidebar.getByText('Unknown server error')).toBeVisible()
```

Comment:
this is because we don't handle MSW errors properly in the UI. One thing we could do, it just occurred to me, is make them look more like real API errors. As in, where right now we just return the Zod error directly as JSON, we could stick it somewhere in an object that looks like `{ "error_code": "InvalidRequest", "message": "..." }`. The Zod error would probably look like garbage stringified, but maybe we put a short version of it in `message` and then log the full thing to the console.

Reply:
https://github.com/oxidecomputer/oxide.ts/issues/228